okhttp/src/main/java/okhttp3/RealCall.kt

```diff
@@ -0,0 +1,214 @@
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.RejectedExecutionException
+import java.util.concurrent.atomic.AtomicInteger
+
+internal class RealCall private constructor(
+  val client: OkHttpClient,
+  /** The application's original request unadulterated by redirects or auth headers.  */
+  val originalRequest: Request,
+  // TODO(egorand): Remove @JvmField once Dispatcher.java is converted to Kotlin
+  @JvmField val forWebSocket: Boolean
+) : Call {
+  /**
+   * There is a cycle between the [Call] and [Transmitter] that makes this awkward.
+   * This is set after immediately after creating the call instance.
+   */
+  private lateinit var transmitter: Transmitter
+
+  // Guarded by this.
+  @get:Synchronized override var isExecuted: Boolean = false
+
+  override val isCanceled: Boolean
+    get() = transmitter.isCanceled
+
+  override fun request(): Request = originalRequest
+
+  @Throws(IOException::class)
+  override fun execute(): Response {
+    synchronized(this) {
+      check(!isExecuted) { "Already Executed" }
+      isExecuted = true
+    }
+    transmitter.timeoutEnter()
+    transmitter.callStart()
+    try {
+      client.dispatcher().executed(this)
+      return getResponseWithInterceptorChain()
+    } finally {
+      client.dispatcher().finished(this)
+    }
+  }
+
+  override fun enqueue(responseCallback: Callback) {
+    synchronized(this) {
+      check(!isExecuted) { "Already Executed" }
+      isExecuted = true
+    }
+    transmitter.callStart()
+    client.dispatcher().enqueue(AsyncCall(responseCallback))
+  }
+
+  override fun cancel() {
+    transmitter.cancel()
+  }
+
+  override fun timeout(): Timeout = transmitter.timeout()
+
+  @SuppressWarnings("CloneDoesntCallSuperClone") // We are a final type & this saves clearing state.
+  override fun clone(): RealCall {
+    return RealCall.newRealCall(client, originalRequest, forWebSocket)
+  }
+
+  internal inner class AsyncCall(
+    private val responseCallback: Callback
+  ) : NamedRunnable("OkHttp %s", redactedUrl()) {
+    @Volatile private var callsPerHost = AtomicInteger(0)
+
+    fun callsPerHost(): AtomicInteger = callsPerHost
+
+    fun reuseCallsPerHostFrom(other: AsyncCall) {
+      this.callsPerHost = other.callsPerHost
+    }
+
+    fun host(): String = originalRequest.url().host()
+
+    fun request(): Request = originalRequest
+
+    fun get(): RealCall = this@RealCall
+
+    /**
+     * Attempt to enqueue this async call on `executorService`. This will attempt to clean up
+     * if the executor has been shut down by reporting the call as failed.
+     */
+    fun executeOn(executorService: ExecutorService) {
+      assert(!Thread.holdsLock(client.dispatcher()))
+      var success = false
+      try {
+        executorService.execute(this)
+        success = true
+      } catch (e: RejectedExecutionException) {
+        val ioException = InterruptedIOException("executor rejected")
+        ioException.initCause(e)
```

Comment:
We should file a JDK bug to get this added to constructor.

Reply:
Good call, wanna do that? Guess you've got more context.

Reply:
Uh I can try! We'll basically be dead before we can use it, but still worth trying to make the world a better place.